File: 백준/Silver/2156. 포도주 시식/포도주 시식.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {

	public static void main(String[] args) throws IOException {

		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

		int n = Integer.parseInt(br.readLine());
		int[] wines = new int[n + 1];
		int[] memo = new int[n + 1];

		for (int i = 1; i <= n; i++) {
			wines[i] = Integer.parseInt(br.readLine());
		}

		memo[1] = wines[1];

		if (n >= 2) {
			memo[2] = memo[1] + wines[2];
		}

		for (int i = 3; i <= n; i++) {
			memo[i] = Math.max(memo[i - 1], Math.max(memo[i - 2] + wines[i], memo[i - 3] + wines[i - 1] + wines[i]));
		}

		System.out.println(memo[n]);
	}
}

File: 백준/Silver/2156. 포도주 시식/포도주 시식.py
import sys

n = int(sys.stdin.readline().rstrip())
wines = [0 for _ in range(n + 1)]
memo = [0 for _ in range(n + 1)]

for i in range(1, n + 1):
    wines[i] = int(sys.stdin.readline().rstrip())

memo[1] = wines[1]

if n >= 2:
    memo[2] = memo[1] + wines[2]

for i in range(3, n + 1):
    memo[i] = max(memo[i - 1], memo[i - 2] + wines[i], memo[i - 3] + wines[i - 1] + wines[i])

print(memo[n])
